pkg/cache/stale.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2021 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Functions in this file solve the problem of duplicate reconciles when the cached
+// client hasn't received the latest resource and feeds a stale resource to controller.
+// We will reject reconciling a resource if its generation is older than one we've
+// successfully reconciled.
+
+// UIDToGenerationMap keeps track of highest observed generation of a particular resource UID
+type UIDToGenerationMap struct {
+	mutex                   sync.RWMutex
+	resourceUIDtoGeneration map[types.UID]int64
```

Comment:
If we have this struct per controller, I wonder if we can get rid of UID as it is not directly used. Can the same thing be achieved by keeping track of 1 generation instead of full history? So the logic would look like - 

```golang
func (u *UIDToGenerationMap) IsCacheStale(generation int64) bool {
	lastGeneration := u.getLastGeneration()
	// Resource should be reconciled if we've never seen it
	if generation == -1 {
		return false
	}
	// Resource is stale if we've reconciled a newer generation
	if generation <= lastGeneration {
		return true
	}
	return false
}
```

Reply:
2 things:

1) I think I should have used resourceVersion and not generation, since that's available everywhere and generation is not .
2) I think only keeping track of a single resourceVersion would mean that we ignore any CR that wasn't the latest one me made an update to, so our reconciler would be broken.